src/state.cairo

```diff
@@ -70,77 +71,89 @@ pub struct Transaction {
+///
+/// There are four possible cases:
+///   1. Coinbase input that does not spend any outputs (zero txid)
+///   2. Input that spends an output created within the same block (cached)
+///   3. Input that spends a coinbase output
+///   4. Input that spends an output from a past block
+///
+/// For (1) we don't need to add extra constraints, because meta fields are not used.
+/// For (2) we need to check that the referenced output is indeed cached:
+///     * Calculate cache key by hashing (txid, vout, data)
+///     * Check if the key is present in the cache
+///     * Remove item from the cache
+/// For (3) we need to check that the referenced output is in the utreexo accumulator:
+///     * Calculate utreexo leaf hash from (txid, vout, data, block_height)
+///     * Verify inclusion proof (either individual or batched) against the roots
+///     * Delete the leaf from the accumulator
+/// For (4) we need to additionally check if the coinbase output is older than 100 blocks
+///
+/// IMPORTANT:
+///     * Utreexo proofs can be verified at any point of block validation because accumulator
+///       is not changing until the end of the block;
+///     * Cache lookups MUST be done in a sequential order, i.e. transactions are validated
+///       one by one, first inputs then outputs. Output validation might put something to the
+///       cache while input validation might remove an item, thus it's important to maintain
+///       the order.
 #[derive(Drop, Copy)]
 pub struct OutPoint {
     /// The hash of the referenced transaction.
     pub txid: Hash,
     /// The index of the specific output in the transaction.
     pub vout: u32,
-    /// The index of output in the utreexo set (meta field).
-    pub txo_index: u64,
-    // Amount calculated with the txid and vout
-    pub amount: u64
+    /// Referenced output data (meta field).
+    /// Must be set to default for coinbase inputs.
+    pub data: TxOut,
+    /// The height of the block that contains this output (meta field).
+    /// Must be set to None for all inputs except for the ones that spend coinbase outputs.
+    pub block_height: Option<u32>,
 }
 
-/// Represents an output of a transaction.
+/// Output of a transaction.
 /// https://learnmeabitcoin.com/technical/transaction/output/
+///
+/// Upon processing (validating) an output one of three actions must be taken:
+///     - Add output with some extra info (see [OutPoint]) to the Utreexo accumulator
+///     - Add output to the cache in case it is going to be spent in the same block
+///     - Do nothing in case of a provably unspendable output
+///
+/// Read more: https://en.bitcoin.it/wiki/Script#Provably_Unspendable/Prunable_Outputs
 #[derive(Drop, Copy)]
 pub struct TxOut {
     /// The value of the output in satoshis.
+    /// The protocol allows zero outputs.
     pub value: u64,
     /// The spending script (aka locking code) for this output.
     pub pk_script: @ByteArray,
```

Comment:
Usually in the transaction output we need to also supply the script byte length. It's OK to omit it here?

Reply:
Yes, script length can be obtained from pk_script byte array